// storybook/dist/cli/index.js
import CJS_COMPAT_NODE_URL_qr9p03ym1e9 from 'node:url';
import CJS_COMPAT_NODE_PATH_qr9p03ym1e9 from 'node:path';
import CJS_COMPAT_NODE_MODULE_qr9p03ym1e9 from "node:module";

var __filename = CJS_COMPAT_NODE_URL_qr9p03ym1e9.fileURLToPath(import.meta.url);
var __dirname = CJS_COMPAT_NODE_PATH_qr9p03ym1e9.dirname(__filename);
var require = CJS_COMPAT_NODE_MODULE_qr9p03ym1e9.createRequire(import.meta.url);

// ------------------------------------------------------------
// end of CJS compatibility banner, injected by Storybook's esbuild configuration
// ------------------------------------------------------------
import {
  ANGULAR_JSON_PATH,
  AddonVitestService,
  AngularJSON,
  ProjectType,
  SUPPORTED_ESLINT_EXTENSIONS,
  addToDevDependenciesIfNotPresent,
  adjustTemplate,
  cliStoriesTargetPath,
  coerceSemver,
  configureEslintPlugin,
  configureFlatConfig,
  copyTemplate,
  copyTemplateFiles,
  editJsonText,
  extractEslintInfo,
  findEslintFile,
  getBabelDependencies,
  getRendererDir,
  getVersionSafe,
  hasStorybookDependencies,
  isStorybookTarget,
  normalizeExtends,
  readFileAsJson,
  suggestESLintPlugin,
  writeFileAsJson
} from "../_node-chunks/chunk-B4IABTQ3.js";
import "../_node-chunks/chunk-I6XXMEZR.js";
import {
  detectIncompatiblePackageVersions,
  detectLanguage,
  getStorybookData,
  getWorkingDir
} from "../_node-chunks/chunk-A4WJ7EKJ.js";
import {
  detectPnp
} from "../_node-chunks/chunk-Z7UBLSJR.js";
import "../_node-chunks/chunk-VEVLFFRF.js";
import {
  Settings,
  _clearGlobalSettings,
  globalSettings
} from "../_node-chunks/chunk-6ULFHL6K.js";
import "../_node-chunks/chunk-EEJDL3CO.js";
import "../_node-chunks/chunk-WWUYXBZF.js";
import "../_node-chunks/chunk-PHW34MLA.js";
import "../_node-chunks/chunk-COYLOOWU.js";
import "../_node-chunks/chunk-7ECT64CN.js";
import "../_node-chunks/chunk-KLIPUMKK.js";
import "../_node-chunks/chunk-VV5PPQIL.js";
import "../_node-chunks/chunk-EOPS5YMG.js";
import "../_node-chunks/chunk-TLXKE3II.js";
export {
  ANGULAR_JSON_PATH,
  AddonVitestService,
  AngularJSON,
  ProjectType,
  SUPPORTED_ESLINT_EXTENSIONS,
  Settings,
  _clearGlobalSettings,
  addToDevDependenciesIfNotPresent,
  adjustTemplate,
  cliStoriesTargetPath,
  coerceSemver,
  configureEslintPlugin,
  configureFlatConfig,
  copyTemplate,
  copyTemplateFiles,
  detectIncompatiblePackageVersions,
  detectLanguage,
  detectPnp,
  editJsonText,
  extractEslintInfo,
  findEslintFile,
  getBabelDependencies,
  getRendererDir,
  getStorybookData,
  getVersionSafe,
  getWorkingDir,
  globalSettings,
  hasStorybookDependencies,
  isStorybookTarget,
  normalizeExtends,
  readFileAsJson,
  suggestESLintPlugin,
  writeFileAsJson
};
