// storybook/dist/node-logger/index.js
import CJS_COMPAT_NODE_URL_qr9p03ym1e9 from 'node:url';
import CJS_COMPAT_NODE_PATH_qr9p03ym1e9 from 'node:path';
import CJS_COMPAT_NODE_MODULE_qr9p03ym1e9 from "node:module";

var __filename = CJS_COMPAT_NODE_URL_qr9p03ym1e9.fileURLToPath(import.meta.url);
var __dirname = CJS_COMPAT_NODE_PATH_qr9p03ym1e9.dirname(__filename);
var require = CJS_COMPAT_NODE_MODULE_qr9p03ym1e9.createRequire(import.meta.url);

// ------------------------------------------------------------
// end of CJS compatibility banner, injected by Storybook's esbuild configuration
// ------------------------------------------------------------
import {
  CLI_COLORS,
  ConsoleLogger,
  StyledConsoleLogger,
  colors,
  createHyperlink,
  deprecate,
  import_npmlog,
  logTracker,
  logger,
  once,
  prompt,
  protectUrls
} from "../_node-chunks/chunk-LKXCSQ7V.js";
import "../_node-chunks/chunk-VV5PPQIL.js";
import "../_node-chunks/chunk-TLXKE3II.js";
var export_instance = import_npmlog.default;
export {
  CLI_COLORS,
  ConsoleLogger,
  StyledConsoleLogger,
  colors,
  createHyperlink,
  deprecate,
  export_instance as instance,
  logTracker,
  logger,
  once,
  prompt,
  protectUrls
};
