// storybook/dist/babel/index.js
import CJS_COMPAT_NODE_URL_qr9p03ym1e9 from 'node:url';
import CJS_COMPAT_NODE_PATH_qr9p03ym1e9 from 'node:path';
import CJS_COMPAT_NODE_MODULE_qr9p03ym1e9 from "node:module";

var __filename = CJS_COMPAT_NODE_URL_qr9p03ym1e9.fileURLToPath(import.meta.url);
var __dirname = CJS_COMPAT_NODE_PATH_qr9p03ym1e9.dirname(__filename);
var require = CJS_COMPAT_NODE_MODULE_qr9p03ym1e9.createRequire(import.meta.url);

// ------------------------------------------------------------
// end of CJS compatibility banner, injected by Storybook's esbuild configuration
// ------------------------------------------------------------
import {
  BabelFileClass,
  babelParse,
  babelParseExpression,
  babelPrint,
  canUpdateVitestConfigFile,
  canUpdateVitestWorkspaceFile,
  core,
  generate,
  getConfigObjectFromMergeArg,
  getEffectiveMergeConfigCall,
  getTargetConfigObject,
  import_core,
  isDefineConfigLike,
  isImportedDefineConfigLikeIdentifier,
  parser,
  parserOptions,
  recast,
  resolveExpression,
  traverse,
  types,
  unwrapTSExpression
} from "../_node-chunks/chunk-I6XXMEZR.js";
import "../_node-chunks/chunk-VEVLFFRF.js";
import "../_node-chunks/chunk-VV5PPQIL.js";
import "../_node-chunks/chunk-TLXKE3II.js";
var export_transformSync = import_core.transformSync;
export {
  BabelFileClass,
  babelParse,
  babelParseExpression,
  babelPrint,
  canUpdateVitestConfigFile,
  canUpdateVitestWorkspaceFile,
  core,
  generate,
  getConfigObjectFromMergeArg,
  getEffectiveMergeConfigCall,
  getTargetConfigObject,
  isDefineConfigLike,
  isImportedDefineConfigLikeIdentifier,
  parser,
  parserOptions,
  recast,
  resolveExpression,
  export_transformSync as transformSync,
  traverse,
  types,
  unwrapTSExpression
};
