// storybook/dist/server-errors.js
import CJS_COMPAT_NODE_URL_qr9p03ym1e9 from 'node:url';
import CJS_COMPAT_NODE_PATH_qr9p03ym1e9 from 'node:path';
import CJS_COMPAT_NODE_MODULE_qr9p03ym1e9 from "node:module";

var __filename = CJS_COMPAT_NODE_URL_qr9p03ym1e9.fileURLToPath(import.meta.url);
var __dirname = CJS_COMPAT_NODE_PATH_qr9p03ym1e9.dirname(__filename);
var require = CJS_COMPAT_NODE_MODULE_qr9p03ym1e9.createRequire(import.meta.url);

// ------------------------------------------------------------
// end of CJS compatibility banner, injected by Storybook's esbuild configuration
// ------------------------------------------------------------
import {
  AddonVitestPostinstallConfigUpdateError,
  AddonVitestPostinstallError,
  AddonVitestPostinstallFailedAddonA11yError,
  AddonVitestPostinstallPrerequisiteCheckError,
  AddonVitestPostinstallWorkspaceUpdateError,
  AngularLegacyBuildOptionsError,
  AutomigrateError,
  Category,
  CommonJsConfigNotSupportedError,
  ConflictingStaticDirConfigError,
  CouldNotEvaluateFrameworkError,
  CriticalPresetLoadError,
  FindPackageVersionsError,
  GenerateNewProjectOnInitError,
  GoogleFontsDownloadError,
  GoogleFontsLoadingError,
  IncompatiblePostCssConfigError,
  InvalidFrameworkNameError,
  InvalidStoriesEntryError,
  MainFileEvaluationError,
  MainFileMissingError,
  MinimumReleaseAgeHandledError,
  MissingAngularJsonError,
  MissingBuilderError,
  MissingFrameworkFieldError,
  NoFreePortError,
  NoMatchingExportError,
  NoStatsForViteDevError,
  NxProjectDetectedError,
  OpenServiceAsyncSchemaError,
  OpenServiceDocgenMissingComponentError,
  OpenServiceDuplicateRegistrationError,
  OpenServiceInvalidStaticPathError,
  OpenServiceLoadedDrainExceededError,
  OpenServiceMissingChannelError,
  OpenServiceMissingServiceError,
  OpenServiceRemoteCommandDisconnectedError,
  OpenServiceRemoteCommandUnhandledError,
  OpenServiceUnimplementedOperationError,
  OpenServiceValidationError,
  SavingGlobalSettingsFileError,
  StatusTypeIdMismatchError,
  StorybookError,
  SvelteViteWithSvelteKitError,
  UpgradeStorybookToLowerVersionError,
  UpgradeStorybookUnknownCurrentVersionError,
  ViteModuleGraphSubscriptionError,
  WebpackCompilationError,
  WebpackInvocationError,
  WebpackMissingStatsError
} from "./_node-chunks/chunk-7ECT64CN.js";
import "./_node-chunks/chunk-KLIPUMKK.js";
import "./_node-chunks/chunk-VV5PPQIL.js";
import "./_node-chunks/chunk-TLXKE3II.js";
export {
  AddonVitestPostinstallConfigUpdateError,
  AddonVitestPostinstallError,
  AddonVitestPostinstallFailedAddonA11yError,
  AddonVitestPostinstallPrerequisiteCheckError,
  AddonVitestPostinstallWorkspaceUpdateError,
  AngularLegacyBuildOptionsError,
  AutomigrateError,
  Category,
  CommonJsConfigNotSupportedError,
  ConflictingStaticDirConfigError,
  CouldNotEvaluateFrameworkError,
  CriticalPresetLoadError,
  FindPackageVersionsError,
  GenerateNewProjectOnInitError,
  GoogleFontsDownloadError,
  GoogleFontsLoadingError,
  IncompatiblePostCssConfigError,
  InvalidFrameworkNameError,
  InvalidStoriesEntryError,
  MainFileEvaluationError,
  MainFileMissingError,
  MinimumReleaseAgeHandledError,
  MissingAngularJsonError,
  MissingBuilderError,
  MissingFrameworkFieldError,
  NoFreePortError,
  NoMatchingExportError,
  NoStatsForViteDevError,
  NxProjectDetectedError,
  OpenServiceAsyncSchemaError,
  OpenServiceDocgenMissingComponentError,
  OpenServiceDuplicateRegistrationError,
  OpenServiceInvalidStaticPathError,
  OpenServiceLoadedDrainExceededError,
  OpenServiceMissingChannelError,
  OpenServiceMissingServiceError,
  OpenServiceRemoteCommandDisconnectedError,
  OpenServiceRemoteCommandUnhandledError,
  OpenServiceUnimplementedOperationError,
  OpenServiceValidationError,
  SavingGlobalSettingsFileError,
  StatusTypeIdMismatchError,
  StorybookError,
  SvelteViteWithSvelteKitError,
  UpgradeStorybookToLowerVersionError,
  UpgradeStorybookUnknownCurrentVersionError,
  ViteModuleGraphSubscriptionError,
  WebpackCompilationError,
  WebpackInvocationError,
  WebpackMissingStatsError
};
